### ISSUE TITLE
In search of a basic validation example

### ISSUE THREAD

**Issue by davidjconnelly** [strongloop/loopback#3724]
Hello,

This is the first time I've ever posted on Github (yes, really!) and I hope I haven't broken any rules by asking a stupid question.  

I'm currently trying to find a **basic and complete example** of how validation works.  All I'm really trying to figure out is how the basic structure/flow would work.

The best example I can find is from raymondfeng on this page:
https://github.com/strongloop/loopback/issues/3676 
(about 3/4 of the way down the page - comment posted 3rd November).

However, what's really confusing me is how   _"validations": []_,  (on the JSON files) fits into the picture.

What is particularly scary for me at this time is that I've started a tutorial series on YouTube, covering Loopback.  I've recorded 15 videos so far and I'm really enjoying learning about Loopback.  As far as I can tell, Loopback is the single best web development technology that I've seen in my entire career.  I'm a huge fan.  However, I'm eager to represent it well and get things right - especially on YouTube.

So, if anyone could give me any advice on best practices for validation then I'd be really grateful.  All I'm really looking for is the most basic example possible (say, make sure a person's age is > 18).

HERE ARE SOME SPECIFIC THINGS THAT I CANNOT FIGURE OUT:

* Let's say we define how validations should work on our model.js file.  Does that negate the need for using validations[] on our JSON file?

* Is the answer to the above is 'yes', does that mean that we can delete the line on the JSON file that says validations[] ?

* Is it a good idea to use **before hooks** for validation or would a '**before save**' vibe be considered a better practice?

* If a person submits a form and there are (let's say) three different and distinct validation errors, do you get Loopback to present a full list (array) of all of the errors OR do you simply throw an error the moment one error appears (therefore, only ever giving an error msg for one error)?

I apologise if these questions have been covered elsewhere.  Let me assure you that I have put a solid two hours into this and I just cannot find any answers.

Many thanks!

**Comment by jannyHou** [comment]
Hi @davidjconnelly welcome to the git community :) Have you looked into our documentation for "validating model data"?

Link: https://loopback.io/doc/en/lb2/Validating-model-data.html

I can provide more help if it doesn't answer your questions.

**Comment by davidjconnelly** [comment]
Hi.  Thank you for your message.  

The answer is yes.  I did see it.  Very respectfully, I'd have to say that I found the documentation to be less than perfect.

In my original post, I had typed out some very specific questions that I have regarding validation.  One or two of those questions have now been answered, pretty much because time seems to have the effect of solving problems.  Nevertheless, I do have a few outstanding questions and I'm slightly disappointed that I'm now effectively being told to go and read the manual.  That type of response, sugar-coated or not, is precisely the reason why people get put off leaving comments and, eventually things like [this](https://kohanaframework.org/) happen.

Asking a question here may have been a mistake.  

Nevertheless, I appreciate your response and if any Loopback makers should ever read this then I'd like to just congratulate you on producing some a great technology.  It's certainly a fabulous thing and I'd even call it a breakthrough for web developers.  So, thank you for that.

Regards,

DC
